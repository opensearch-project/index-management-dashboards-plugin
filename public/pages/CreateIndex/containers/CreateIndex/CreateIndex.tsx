/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component, useContext } from "react";
import { EuiSpacer, EuiTitle } from "@elastic/eui";
import { RouteComponentProps } from "react-router-dom";
import IndexForm from "../IndexForm";
import { BREADCRUMBS, IndicesUpdateMode, ROUTES } from "../../../../utils/constants";
import { CoreServicesContext } from "../../../../components/core_services";
import { DataSourceMenuContext, DataSourceMenuProperties } from "../../../../services/DataSourceMenuContext";
import { useUpdateUrlWithDataSourceProperties } from "../../../../components/MDSEnabledComponent";
import { getUISettings } from "../../../../services/Services";

interface CreateIndexPropsBase extends RouteComponentProps<{ index?: string; mode?: IndicesUpdateMode }> {
  isEdit?: boolean;
  useUpdatedUX?: boolean;
}

interface CreateIndexProps extends CreateIndexPropsBase, DataSourceMenuProperties {}

export class CreateIndex extends Component<CreateIndexProps> {
  static contextType = CoreServicesContext;

  get index() {
    return this.props.match.params.index;
  }

  get isEdit() {
    return this.props.match.params.index !== undefined;
  }

  componentDidMount = async (): Promise<void> => {
    const isEdit = this.isEdit;
    const breadCrumbs = this.props.useUpdatedUX
      ? [BREADCRUMBS.INDICES, isEdit ? BREADCRUMBS.EDIT_INDEX : BREADCRUMBS.CREATE_INDEX]
      : [BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.INDICES, isEdit ? BREADCRUMBS.EDIT_INDEX : BREADCRUMBS.CREATE_INDEX];
    this.context.chrome.setBreadcrumbs(breadCrumbs);
  };

  onCancel = (): void => {
    this.props.history.push(ROUTES.INDICES);
  };

  render() {
    const isEdit = this.isEdit;

    return this.props.useUpdatedUX ? (
      <div style={{ padding: "0px 0px" }}>
        <IndexForm
          index={this.index}
          mode={this.props.match.params.mode}
          onCancel={this.onCancel}
          onSubmitSuccess={() => this.props.history.push(ROUTES.INDICES)}
          dataSourceId={this.props.dataSourceId}
          useUpdatedUX={this.props.useUpdatedUX}
        />
      </div>
    ) : (
      <div style={{ padding: "0px 50px" }}>
        <EuiTitle size="l">
          <h1>{isEdit ? "Edit" : "Create"} index</h1>
        </EuiTitle>
        <EuiSpacer />
        <IndexForm
          index={this.index}
          mode={this.props.match.params.mode}
          onCancel={this.onCancel}
          onSubmitSuccess={() => this.props.history.push(ROUTES.INDICES)}
          dataSourceId={this.props.dataSourceId}
          useUpdatedUX={this.props.useUpdatedUX}
        />
      </div>
    );
  }
}

export default function (props: CreateIndexPropsBase) {
  const dataSourceMenuProperties = useContext(DataSourceMenuContext);
  useUpdateUrlWithDataSourceProperties();
  const uiSettings = getUISettings();
  const useUpdatedUX = uiSettings.get("home:useNewHomePage");
  return <CreateIndex {...props} {...dataSourceMenuProperties} useUpdatedUX={useUpdatedUX} />;
}
