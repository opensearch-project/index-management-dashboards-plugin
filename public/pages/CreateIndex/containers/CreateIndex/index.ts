/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import CreateIndex from "./CreateIndex";

export default CreateIndex;
