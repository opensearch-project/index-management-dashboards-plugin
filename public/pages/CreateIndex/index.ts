/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import CreateIndex from "./containers/CreateIndex";

export default CreateIndex;
