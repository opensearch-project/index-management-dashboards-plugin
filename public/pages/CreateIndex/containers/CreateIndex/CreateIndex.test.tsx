/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { MemoryRouter as Router, Route, RouteComponentProps, Switch } from "react-router-dom";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CoreStart } from "opensearch-dashboards/public";
import CreateIndex from "./CreateIndex";
import { ServicesConsumer, ServicesContext } from "../../../../services";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import { BrowserServices } from "../../../../models/interfaces";
import { ModalProvider, ModalRoot } from "../../../../components/Modal";
import { IndicesUpdateMode, ROUTES } from "../../../../utils/constants";
import { CoreServicesConsumer, CoreServicesContext } from "../../../../components/core_services";

browserServicesMock.commonService.apiCaller = jest.fn(
  async (payload): Promise<any> => {
    switch (payload.endpoint) {
      case "transport.request": {
        if (payload.data?.path?.startsWith("/_index_template/_simulate_index/")) {
          return {
            ok: true,
            response: {
              template: {
                settings: {
                  index: {
                    number_of_replicas: "10",
                  },
                },
              },
            },
          };
        }
      }
      case "indices.create":
        if (payload.data?.index === "bad_index") {
          return {
            ok: false,
            error: "bad_index",
          };
        }

        return {
          ok: true,
          response: {},
        };
        break;
      case "cat.aliases":
        return {
          ok: true,
          response: [
            {
              alias: ".kibana",
              index: ".kibana_1",
              filter: "-",
              is_write_index: "-",
            },
            {
              alias: "2",
              index: "1234",
              filter: "-",
              is_write_index: "-",
            },
          ],
        };
      case "indices.get":
        const payloadIndex = payload.data?.index;
        if (payloadIndex === "bad_index") {
          return {
            ok: false,
            error: "bad_error",
            response: {},
          };
        }

        return {
          ok: true,
          response: {
            [payload.data?.index]: {
              aliases: {
                update_test_1: {},
              },
              mappings: {
                properties: {
                  test_mapping_1: {
                    type: "text",
                  },
                },
              },
              settings: {
                "index.number_of_shards": "1",
                "index.number_of_replicas": "1",
              },
            },
          },
        };
    }
    return {
      ok: true,
      response: {},
    };
  }
);

function renderCreateIndexWithRouter(initialEntries = [ROUTES.CREATE_INDEX] as string[]) {
  return {
    ...render(
      <Router initialEntries={initialEntries}>
        <CoreServicesContext.Provider value={coreServicesMock}>
          <ServicesContext.Provider value={browserServicesMock}>
            <ServicesConsumer>
              {(services: BrowserServices | null) =>
                services && (
                  <CoreServicesConsumer>
                    {(core: CoreStart | null) =>
                      core && (
                        <ModalProvider>
                          <ModalRoot services={services} />
                          <Switch>
                            <Route
                              path={`${ROUTES.CREATE_INDEX}/:index/:mode`}
                              render={(props: RouteComponentProps) => <CreateIndex {...props} commonService={services.commonService} />}
                            />
                            <Route
                              path={`${ROUTES.CREATE_INDEX}/:index`}
                              render={(props: RouteComponentProps) => <CreateIndex {...props} commonService={services.commonService} />}
                            />
                            <Route
                              path={ROUTES.CREATE_INDEX}
                              render={(props: RouteComponentProps) => <CreateIndex {...props} commonService={services.commonService} />}
                            />
                            <Route
                              path={ROUTES.INDICES}
                              render={(props: RouteComponentProps) => <h1>location is: {ROUTES.INDEX_POLICIES}</h1>}
                            />
                          </Switch>
                        </ModalProvider>
                      )
                    }
                  </CoreServicesConsumer>
                )
              }
            </ServicesConsumer>
          </ServicesContext.Provider>
        </CoreServicesContext.Provider>
      </Router>
    ),
  };
}

describe("<CreateIndex /> spec", () => {
  it("renders the create component", async () => {
    const { container } = renderCreateIndexWithRouter();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("show a toast if getIndices gracefully fails", async () => {
    const { getByText } = renderCreateIndexWithRouter([`${ROUTES.CREATE_INDEX}/bad_index`]);

    await waitFor(() => {
      getByText("Update");
    });
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("bad_error");
  });

  it("shows error for index name input when clicking create", async () => {
    const { queryByText, getByText } = renderCreateIndexWithRouter();

    await waitFor(() => getByText("Define index"));

    expect(queryByText("Index name can not be null.")).toBeNull();

    userEvent.click(getByText("Create"));
    await waitFor(() => {
      expect(queryByText("Index name can not be null.")).not.toBeNull();
    });
  });

  it("routes you back to indices and shows a success toast when successfully creating a index", async () => {
    const { getByText, getByPlaceholderText, getByTestId } = renderCreateIndexWithRouter();

    await waitFor(() => {
      getByText("Define index");
    });

    userEvent.type(getByPlaceholderText("Please enter the name for your index"), `some_index`);
    userEvent.click(document.body);
    await waitFor(() => {
      expect(getByTestId("form-name-index.number_of_replicas").querySelector("input")).toHaveAttribute("value", "10");
    });
    userEvent.click(getByText("Create"));
    await waitFor(() => {
      expect(coreServicesMock.notifications.toasts.addSuccess).toHaveBeenCalledWith("some_index has been successfully created.");
    });
  });

  it("shows a danger toast when getting graceful error from create index", async () => {
    const { getByText, getByPlaceholderText } = renderCreateIndexWithRouter();

    await waitFor(() => getByText("Define index"));

    userEvent.type(getByPlaceholderText("Please enter the name for your index"), `bad_index`);
    userEvent.click(getByText("Create"));

    await waitFor(() => {
      expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("bad_index");
    });
  });

  it("it shows detail and does not call any api when nothing modified", async () => {
    const { getByText, getByTestId } = renderCreateIndexWithRouter([`${ROUTES.CREATE_INDEX}/good_index`]);
    await waitFor(() => getByText("Define index"));
    userEvent.click(getByText("Update"));
    await waitFor(() => {});
    userEvent.click(getByTestId("change_diff_confirm-confirm"));

    await waitFor(() => {
      // it shows detail and does not call any api when nothing modified
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(2);
    });
  });

  it("shows detail info and update others", async () => {
    const { getByText, getByTestId, getByTitle } = renderCreateIndexWithRouter([`${ROUTES.CREATE_INDEX}/good_index`]);

    await waitFor(() => getByText("Define index"));

    userEvent.click(getByTitle("update_test_1").querySelector("button") as Element);
    userEvent.type(getByTestId("comboBoxSearchInput"), "test_1{enter}");
    userEvent.type(getByTestId("form-name-index.number_of_replicas").querySelector("input") as Element, "2");
    userEvent.click(getByTestId("create index add field button"));
    await waitFor(() => {});
    await userEvent.clear(getByTestId("mapping-visual-editor-1-field-name"));
    await userEvent.type(getByTestId("mapping-visual-editor-1-field-name"), "test_mapping_2");
    await userEvent.click(document.body);
    userEvent.click(getByText("Update"));
    await waitFor(() => {});
    userEvent.click(getByTestId("change_diff_confirm-confirm"));

    await waitFor(() => {
      expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalledTimes(1);
      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        endpoint: "indices.updateAliases",
        method: "PUT",
        data: {
          body: {
            actions: [
              {
                remove: {
                  index: "good_index",
                  alias: "update_test_1",
                },
              },
              {
                add: {
                  index: "good_index",
                  alias: "test_1",
                },
              },
            ],
          },
        },
      });

      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        endpoint: "indices.putSettings",
        method: "PUT",
        data: {
          index: "good_index",
          flat_settings: true,
          body: {
            "index.number_of_replicas": "12",
          },
        },
      });

      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        endpoint: "indices.putMapping",
        method: "PUT",
        data: {
          index: "good_index",
          body: {
            properties: {
              test_mapping_2: {
                type: "text",
              },
            },
          },
        },
      });
    });
  });

  it("shows detail alias and update alias only", async () => {
    const { getByText, getByTestId, getByTitle } = renderCreateIndexWithRouter([
      `${ROUTES.CREATE_INDEX}/good_index/${IndicesUpdateMode.alias}`,
    ]);

    await waitFor(() => {});

    userEvent.click(getByTitle("update_test_1").querySelector("button") as Element);
    userEvent.type(getByTestId("comboBoxSearchInput"), "test_1{enter}");
    await waitFor(() => {});
    userEvent.click(getByText("Update"));
    await waitFor(() => {});
    userEvent.click(getByTestId("change_diff_confirm-cancel"));
    await waitFor(() => {});
    userEvent.click(getByText("Update"));
    await waitFor(() => {});
    userEvent.click(getByTestId("change_diff_confirm-confirm"));

    await waitFor(() => {
      // shows detail alias and update alias only
      expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalledTimes(1);
      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        endpoint: "indices.updateAliases",
        method: "PUT",
        data: {
          body: {
            actions: [
              {
                remove: {
                  index: "good_index",
                  alias: "update_test_1",
                },
              },
              {
                add: {
                  index: "good_index",
                  alias: "test_1",
                },
              },
            ],
          },
        },
      });
    });
  });

  it("shows detail settings and update settings only", async () => {
    const { getByText, getByTestId } = renderCreateIndexWithRouter([`${ROUTES.CREATE_INDEX}/good_index/${IndicesUpdateMode.mappings}`]);

    await waitFor(() => {});

    userEvent.click(getByTestId("create index add field button"));
    await waitFor(() => {});
    await userEvent.clear(getByTestId("mapping-visual-editor-1-field-name"));
    await userEvent.type(getByTestId("mapping-visual-editor-1-field-name"), "test_mapping_2");
    await userEvent.click(document.body);
    await waitFor(() => {});
    userEvent.click(getByText("Update"));
    await waitFor(() => {});
    userEvent.click(getByTestId("change_diff_confirm-confirm"));

    await waitFor(() => {
      // shows detail settings and update settings only
      expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalledTimes(1);

      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        endpoint: "indices.putMapping",
        method: "PUT",
        data: {
          index: "good_index",
          body: {
            properties: {
              test_mapping_2: {
                type: "text",
              },
            },
          },
        },
      });
    });
  });

  it("shows detail mappings and update mappings only", async () => {
    const { getByText, getByTestId } = renderCreateIndexWithRouter([`${ROUTES.CREATE_INDEX}/good_index/${IndicesUpdateMode.settings}`]);

    await waitFor(() => {});

    userEvent.type(getByTestId("form-name-index.number_of_replicas").querySelector("input") as Element, "2");
    userEvent.click(getByText("Update"));
    await waitFor(() => {});
    userEvent.click(getByTestId("change_diff_confirm-confirm"));

    await waitFor(() => {
      // shows detail mappings and update mappings only
      expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalledTimes(1);

      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        endpoint: "indices.putSettings",
        method: "PUT",
        data: {
          index: "good_index",
          flat_settings: true,
          body: {
            "index.number_of_replicas": "12",
          },
        },
      });
    });
  });

  it("it goes to indices page when click cancel", async () => {
    const { getByText } = renderCreateIndexWithRouter([`${ROUTES.CREATE_INDEX}/good_index`]);
    await waitFor(() => {});
    userEvent.click(getByText("Cancel"));
    await waitFor(() => {
      expect(getByText(`location is: ${ROUTES.INDEX_POLICIES}`)).toBeInTheDocument();
    });
  });
});
