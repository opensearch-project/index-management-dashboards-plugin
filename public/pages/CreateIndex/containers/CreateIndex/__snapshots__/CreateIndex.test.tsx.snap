// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateIndex /> spec renders the create component 1`] = `
<div
  style="padding: 0px 50px;"
>
  <h1
    class="euiTitle euiTitle--large"
  >
    Create
     index
  </h1>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Define index
        </h3>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 0px 10px;"
    >
      <div
        style="padding-left: 10px;"
      >
        <div
          class="euiFormRow"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="some_html_id"
            >
              Index name
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  aria-describedby="some_html_id-help-0"
                  class="euiFieldText"
                  id="some_html_id"
                  placeholder="Please enter the name for your index"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="euiFormHelpText euiFormRow__text"
              id="some_html_id-help-0"
            >
              Please enter the name before moving to other fields
            </div>
          </div>
        </div>
        <div
          class="euiFormRow"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              class="euiFormLabel euiFormRow__label"
              for="some_html_id"
            >
              Index alias  - optional
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              aria-describedby="some_html_id-help-0"
              aria-expanded="false"
              aria-haspopup="listbox"
              class="euiComboBox euiComboBox-isDisabled"
              role="combobox"
            >
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <div
                    class="euiComboBox__inputWrap euiComboBox__inputWrap-isLoading"
                    data-test-subj="comboBoxInput"
                    tabindex="-1"
                  >
                    <p
                      class="euiComboBoxPlaceholder"
                    >
                      Select or create aliases
                    </p>
                    <div
                      class="euiComboBox__input"
                      style="font-size: 14px; display: inline-block;"
                    >
                      <input
                        aria-controls=""
                        data-test-subj="comboBoxSearchInput"
                        disabled=""
                        id="some_html_id"
                        role="textbox"
                        style="box-sizing: content-box; width: 2px;"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                  <div
                    class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <span
                      class="euiLoadingSpinner euiLoadingSpinner--medium"
                    />
                    <button
                      aria-label="Open list of options"
                      class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                      data-test-subj="comboBoxToggleListButton"
                      disabled=""
                      type="button"
                    >
                      EuiIconMock
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFormHelpText euiFormRow__text"
              id="some_html_id-help-0"
            >
              Select existing aliases or specify a new alias
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Index settings
        </h3>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 0px 10px;"
    >
      <div
        style="padding-left: 10px;"
      >
        <div
          class="euiForm"
        >
          <div
            class="euiFormRow"
            data-test-subj="form-name-index.number_of_shards"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                aria-invalid="false"
                class="euiFormLabel euiFormRow__label"
                for="some_html_id"
              >
                Number of shards
              </label>
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormHelpText euiFormRow__text"
                style="padding-top: 0px; padding-bottom: 4px;"
              >
                The number of primary shards in the index. Default is 1.
              </div>
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber"
                    data-meta="Field"
                    disabled=""
                    id="index.number_of_shards"
                    placeholder="The number of primary shards in the index. Default is 1."
                    type="number"
                    value="1"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFormRow"
            data-test-subj="form-name-index.number_of_replicas"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                aria-invalid="false"
                class="euiFormLabel euiFormRow__label"
                for="some_html_id"
              >
                Number of replicas
              </label>
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormHelpText euiFormRow__text"
                style="padding-top: 0px; padding-bottom: 4px;"
              >
                The number of replica shards each primary shard should have.
              </div>
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber"
                    data-meta="Field"
                    disabled=""
                    id="index.number_of_replicas"
                    placeholder="The number of replica shards each primary shard should have."
                    type="number"
                    value="1"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFormRow"
            data-test-subj="form-name-index.refresh_interval"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                aria-invalid="false"
                class="euiFormLabel euiFormRow__label"
                for="some_html_id"
              >
                Refresh interval of index
              </label>
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormHelpText euiFormRow__text"
                style="padding-top: 0px; padding-bottom: 4px;"
              >
                How often the index should refresh, which publishes its most recent changes and makes them available for searching.
              </div>
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldText"
                    data-meta="Field"
                    disabled=""
                    id="index.refresh_interval"
                    placeholder="Can be set to -1 to disable refreshing."
                    type="text"
                    value="1s"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--m"
          />
          <div
            class="euiAccordion"
          >
            <div
              class="euiAccordion__triggerWrapper"
            >
              <button
                aria-controls="accordion_for_create_index_settings"
                aria-expanded="false"
                class="euiAccordion__button"
                id="some_html_id"
                type="button"
              >
                <span
                  class="euiAccordion__iconWrapper"
                >
                  EuiIconMock
                </span>
                <span
                  class="euiIEFlexWrapFix"
                >
                  <h4>
                    Advanced settings
                  </h4>
                </span>
              </button>
            </div>
            <div
              aria-labelledby="some_html_id"
              class="euiAccordion__childWrapper"
              id="accordion_for_create_index_settings"
              role="region"
              tabindex="-1"
            >
              <div>
                <div
                  class=""
                >
                  <div
                    class="euiSpacer euiSpacer--m"
                  />
                  <div
                    class="euiFormRow"
                    data-test-subj="form-name-advanced-settings"
                    id="some_html_id-row"
                  >
                    <div
                      class="euiFormRow__labelWrapper"
                    >
                      <label
                        class="euiFormLabel euiFormRow__label"
                        for="some_html_id"
                      >
                        Specify advanced index settings
                      </label>
                    </div>
                    <div
                      class="euiFormRow__fieldWrapper"
                    >
                      <textarea
                        data-test-subj="json-editor-value-display"
                        readonly=""
                        style="display: none;"
                      >
                        {
  "index.number_of_shards": 1,
  "index.number_of_replicas": 1,
  "index.refresh_interval": "1s"
}
                      </textarea>
                      <div
                        class="euiCodeEditorWrapper"
                        data-test-subj="codeEditorContainer"
                      >
                        <button
                          class="euiCodeEditorKeyboardHint"
                          data-test-subj="codeEditorHint"
                          id="some_html_id"
                        >
                          <p
                            class="euiText"
                          >
                            Press Enter to start editing.
                          </p>
                          <p
                            class="euiText"
                          >
                            When you're done, press Escape to stop editing.
                          </p>
                        </button>
                        <div
                          class=" ace_editor ace-tm"
                          id="some_html_id"
                          style="width: 500px; height: 500px;"
                        >
                          <textarea
                            aria-describedby="some_html_id-help-0"
                            autocapitalize="off"
                            autocorrect="off"
                            class="ace_text-input"
                            spellcheck="false"
                            style="opacity: 0;"
                            tabindex="-1"
                            wrap="off"
                          />
                          <div
                            aria-hidden="true"
                            class="ace_gutter"
                          >
                            <div
                              class="ace_layer ace_gutter-layer ace_folding-enabled"
                            />
                            <div
                              class="ace_gutter-active-line"
                            />
                          </div>
                          <div
                            class="ace_scroller"
                          >
                            <div
                              class="ace_content"
                            >
                              <div
                                class="ace_layer ace_print-margin-layer"
                              >
                                <div
                                  class="ace_print-margin"
                                  style="left: 4px; visibility: visible;"
                                />
                              </div>
                              <div
                                class="ace_layer ace_marker-layer"
                              />
                              <div
                                class="ace_layer ace_text-layer"
                                style="padding: 0px 4px;"
                              />
                              <div
                                class="ace_layer ace_marker-layer"
                              />
                              <div
                                class="ace_layer ace_cursor-layer ace_hidden-cursors"
                              >
                                <div
                                  class="ace_cursor"
                                />
                              </div>
                            </div>
                          </div>
                          <div
                            class="ace_scrollbar ace_scrollbar-v"
                            style="display: none; width: 20px;"
                          >
                            <div
                              class="ace_scrollbar-inner"
                              style="width: 20px;"
                            />
                          </div>
                          <div
                            class="ace_scrollbar ace_scrollbar-h"
                            style="display: none; height: 20px;"
                          >
                            <div
                              class="ace_scrollbar-inner"
                              style="height: 20px;"
                            />
                          </div>
                          <div
                            style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: hidden;"
                          >
                            <div
                              style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
                            />
                            <div
                              style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
                            >
                              XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="euiFormHelpText euiFormRow__text"
                        id="some_html_id-help-0"
                      >
                        Specify a comma-delimited list of settings.
                        <a
                          class="euiLink euiLink--primary"
                          href="https://opensearch.org/docs/latest/api-reference/index-apis/create-index#index-settings"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          View index settings
                          EuiIconMock
                          <span
                            class="euiScreenReaderOnly"
                          >
                            (opens in a new tab or window)
                          </span>
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Index mappings - optional
        </h3>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 0px 10px;"
    >
      <div
        style="padding-left: 10px;"
      >
        <div
          class="euiFormRow euiFormRow--fullWidth"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiSpacer euiSpacer--l"
            />
            <fieldset
              class="euiButtonGroup euiButtonGroup--small euiButtonGroup--text"
            >
              <legend
                class="euiScreenReaderOnly"
              >
                Editor Type
              </legend>
              <div
                class="euiButtonGroup__buttons"
              >
                <label
                  class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small euiButtonGroupButton--fill euiButtonGroupButton-isSelected"
                  data-test-subj="editor-type-visual-editor"
                  for="some_html_id"
                >
                  <span
                    class="euiButtonContent euiButton__content"
                  >
                    <span
                      class="euiButton__text euiButtonGroupButton__textShift"
                      data-text="Visual Editor"
                      title="Visual Editor"
                    >
                      <input
                        checked=""
                        class="euiScreenReaderOnly"
                        data-test-subj="VISUAL"
                        id="some_html_id"
                        name="some_html_id"
                        type="radio"
                        value=""
                      />
                      Visual Editor
                    </span>
                  </span>
                </label>
                <label
                  class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small"
                  data-test-subj="editor-type-json-editor"
                  for="some_html_id"
                >
                  <span
                    class="euiButtonContent euiButton__content"
                  >
                    <span
                      class="euiButton__text euiButtonGroupButton__textShift"
                      data-text="JSON Editor"
                      title="JSON Editor"
                    >
                      <input
                        class="euiScreenReaderOnly"
                        data-test-subj="JSON"
                        id="some_html_id"
                        name="some_html_id"
                        type="radio"
                        value=""
                      />
                      JSON Editor
                    </span>
                  </span>
                </label>
              </div>
            </fieldset>
            <div
              class="euiSpacer euiSpacer--l"
            />
            <p>
              You have no field mappings.
            </p>
            <div
              class="euiSpacer euiSpacer--l"
            />
            <button
              class="euiButton euiButton--primary"
              data-test-subj="create index add field button"
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Add a field
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButtonEmpty euiButtonEmpty--primary"
        data-test-subj="createIndexCancelButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            Cancel
          </span>
        </span>
      </button>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButton euiButton--primary euiButton--fill"
        data-test-subj="createIndexCreateButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            Create
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
